### DISCUSSION TITLE
linking issues

### DISCUSSION THREAD

**Discussion by cwebdesign** [Genivia/RE-flex#206] [Q&A]
I installed RE-flex on macos using homebrew. I use g++-14 from homebrew.
I'm trying to link in the generated code in my app with the reflex library, but I get errors:
Undefined symbols for architecture arm64:
  "std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>::at(unsigned long) const", referenced from:
      reflex::Pattern::graph_dfa(reflex::Pattern::DFA::State const*) const in libreflex.a(libreflex_a-pattern.o)
      reflex::Pattern::gencode_dfa(reflex::Pattern::DFA::State const*) const in libreflex.a(libreflex_a-pattern.o)
      reflex::Pattern::export_code() const in libreflex.a(libreflex_a-pattern.o)
  "std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>::find(char, unsigned long) const", referenced from:
      reflex::Pattern::init_options(char const*) in libreflex.a(libreflex_a-pattern.o)
      reflex::Pattern::parse4(bool, unsigned int&, std::__1::vector<reflex::Pattern::Position, std::__1::allocator<reflex::Pattern::Position>>&, std::__1::vector<reflex::Pattern::Position, std::__1::allocator<reflex::Pattern::Position>>&, bool&, std::__1::map<reflex::Pattern::Position, std::__1::vector<reflex::Pattern::Position, std::__1::allocator<reflex::Pattern::Position>>, std::__1::less<reflex::Pattern::Position>, std::__1::allocator<std::__1::pair<reflex::Pattern::Position const, std::__1::vector<reflex::Pattern::Position, std::__1::allocator<reflex::Pattern::Position>>>>>&, unsigned char&, std::__1::vector<reflex::Pattern::Position, std::__1::allocator<reflex::Pattern::Position>>&, reflex::ORanges<unsigned int>*, reflex::ORanges<unsigned int>&, unsigned short&) in libreflex.a(libreflex_a-pattern.o)
      reflex::Pattern::compile_list(unsigned int, reflex::Pattern::Chars&, reflex::ORanges<unsigned int> const*) const in libreflex.a(libreflex_a-pattern.o)
  "std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>::compare(unsigned long, unsigned long, char const*) const", referenced from:
      reflex::Pattern::parse_esc(unsigned int&, reflex::Pattern::Chars*) const in libreflex.a(libreflex_a-pattern.o)
      reflex::Pattern::compile_list(unsigned int, reflex::Pattern::Chars&, reflex::ORanges<unsigned int> const*) const in libreflex.a(libreflex_a-pattern.o)
      reflex::Pattern::graph_dfa(reflex::Pattern::DFA::State const*) const in libreflex.a(libreflex_a-pattern.o)
      reflex::Pattern::gencode_dfa(reflex::Pattern::DFA::State const*) const in libreflex.a(libreflex_a-pattern.o)
      reflex::Pattern::export_code() const in libreflex.a(libreflex_a-pattern.o)
  "std::__1::locale::use_facet(std::__1::locale::id&) const", referenced from:
      std::__1::basic_ostream<char, std::__1::char_traits<char>>& std::__1::__put_character_sequence<char, std::__1::char_traits<char>>(std::__1::basic_ostream<char, std::__1::char_traits<char>>&, char const*, unsigned long) in libreflex.a(libreflex_a-pattern.o)
  "std::__1::ios_base::getloc() const", referenced from:
      std::__1::basic_ostream<char, std::__1::char_traits<char>>& std::__1::__put_character_sequence<char, std::__1::char_traits<char>>(std::__1::basic_ostream<char, std::__1::char_traits<char>>&, char const*, unsigned long) in libreflex.a(libreflex_a-pattern.o)
  "std::runtime_error::runtime_error(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&)", referenced from:
      reflex::regex_error::regex_error(int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, unsigned long) in libreflex.a(libreflex_a-pattern.o)
  "std::bad_array_new_length::bad_array_new_length()", referenced from:
      std::__throw_bad_array_new_length[abi:v15006]() in libreflex.a(libreflex_a-matcher.o)
      std::__throw_bad_array_new_length[abi:v15006]() in libreflex.a(libreflex_a-pattern.o)
  "std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>::append(char const*)", referenced from:
      reflex::regex_error::regex_error_message(char const*, char const*, unsigned long) in libreflex.a(libreflex_a-error.o)
  "std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>::append(char const*, unsigned long)", referenced from:
      reflex::regex_error::regex_error_message(char const*, char const*, unsigned long) in libreflex.a(libreflex_a-error.o)
  "std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>::append(unsigned long, char)", referenced from:
      reflex::regex_error::regex_error_message(char const*, char const*, unsigned long) in libreflex.a(libreflex_a-error.o)
  "std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>::basic_string(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&)", referenced from:
      reflex::Pattern::Pattern(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, char const*) in libreflex.a(libreflex_a-matcher.o)
      reflex::Pattern::operator[](unsigned int) const in libreflex.a(libreflex_a-pattern.o)
      reflex::Pattern::init_options(char const*) in libreflex.a(libreflex_a-pattern.o)
      void std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>>::__push_back_slow_path<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&>(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) in libreflex.a(libreflex_a-pattern.o)
  "std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>::basic_string(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, unsigned long, unsigned long, std::__1::allocator<char> const&)", referenced from:
      reflex::Pattern::operator[](unsigned int) const in libreflex.a(libreflex_a-pattern.o)
      reflex::Pattern::write_namespace_open(__sFILE*) const in libreflex.a(libreflex_a-pattern.o)
      reflex::Pattern::write_namespace_close(__sFILE*) const in libreflex.a(libreflex_a-pattern.o)
  "std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>::operator=(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&)", referenced from:
      reflex::Pattern::init_options(char const*) in libreflex.a(libreflex_a-pattern.o)
  "std::__1::basic_istream<char, std::__1::char_traits<char>>::get()", referenced from:
      reflex::Input::get(char*, unsigned long) in libreflex.a(libreflex_a-matcher.o)
  "std::__1::basic_istream<char, std::__1::char_traits<char>>::read(char*, long)", referenced from:
      reflex::Input::get(char*, unsigned long) in libreflex.a(libreflex_a-matcher.o)
  "std::__1::basic_istream<char, std::__1::char_traits<char>>::seekg(long long, std::__1::ios_base::seekdir)", referenced from:
      reflex::Input::istream_size() in libreflex.a(libreflex_a-input.o)
  "std::__1::basic_istream<char, std::__1::char_traits<char>>::tellg()", referenced from:
      reflex::Input::istream_size() in libreflex.a(libreflex_a-input.o)
  "std::__1::basic_ostream<char, std::__1::char_traits<char>>::sentry::sentry(std::__1::basic_ostream<char, std::__1::char_traits<char>>&)", referenced from:
      std::__1::basic_ostream<char, std::__1::char_traits<char>>& std::__1::__put_character_sequence<char, std::__1::char_traits<char>>(std::__1::basic_ostream<char, std::__1::char_traits<char>>&, char const*, unsigned long) in libreflex.a(libreflex_a-pattern.o)
  "std::__1::basic_ostream<char, std::__1::char_traits<char>>::sentry::~sentry()", referenced from:
      std::__1::basic_ostream<char, std::__1::char_traits<char>>& std::__1::__put_character_sequence<char, std::__1::char_traits<char>>(std::__1::basic_ostream<char, std::__1::char_traits<char>>&, char const*, unsigned long) in libreflex.a(libreflex_a-pattern.o)
  "std::__1::cerr", referenced from:
      reflex::Pattern::error(int, unsigned long) const in libreflex.a(libreflex_a-pattern.o)
  "std::__1::ctype<char>::id", referenced from:
      std::__1::basic_ostream<char, std::__1::char_traits<char>>& std::__1::__put_character_sequence<char, std::__1::char_traits<char>>(std::__1::basic_ostream<char, std::__1::char_traits<char>>&, char const*, unsigned long) in libreflex.a(libreflex_a-pattern.o)
  "std::__1::locale::~locale()", referenced from:
      std::__1::basic_ostream<char, std::__1::char_traits<char>>& std::__1::__put_character_sequence<char, std::__1::char_traits<char>>(std::__1::basic_ostream<char, std::__1::char_traits<char>>&, char const*, unsigned long) in libreflex.a(libreflex_a-pattern.o)
  "std::__1::ios_base::__set_badbit_and_consider_rethrow()", referenced from:
      std::__1::basic_ostream<char, std::__1::char_traits<char>>& std::__1::__put_character_sequence<char, std::__1::char_traits<char>>(std::__1::basic_ostream<char, std::__1::char_traits<char>>&, char const*, unsigned long) in libreflex.a(libreflex_a-pattern.o)
  "std::__1::ios_base::clear(unsigned int)", referenced from:
      std::__1::basic_ostream<char, std::__1::char_traits<char>>& std::__1::__put_character_sequence<char, std::__1::char_traits<char>>(std::__1::basic_ostream<char, std::__1::char_traits<char>>&, char const*, unsigned long) in libreflex.a(libreflex_a-pattern.o)
  "std::bad_alloc::bad_alloc()", referenced from:
      reflex::AbstractMatcher::grow(unsigned long) in libreflex.a(libreflex_a-matcher.o)
      reflex::AbstractMatcher::reset(char const*) in libreflex.a(libreflex_a-matcher.o)
ld: symbol(s) not found for architecture arm64
collect2: error: ld returned 1 exit status

Any ideas on what I could do?
My code links properly on other platforms.

**Comment by genivia-inc** [comment]
My typical development environment is MacOS M1 (arm64) and haven't had any problems compiling and linking with clang. I've verified that `c++ -std=c++14` works fine, where `c++` runs clang++ version 14.0.0.

Check with `nm`:
```console
$  nm -gU lib/libreflex.a 

lib/libreflex.a(convert.o):
0000000000006af4 T __ZN6reflex11regex_errorC1EiPKcm
000000000000a4c4 T __ZN6reflex11regex_errorD0Ev
0000000000006b90 T __ZN6reflex11regex_errorD1Ev
...
lib/libreflex.a(matcher.o):
000000000000c9e8 T __ZN6reflex14PatternMatcherINS_7PatternEE7patternEPKS1_
000000000000ca3c T __ZN6reflex14PatternMatcherINS_7PatternEE7patternEPKc
000000000000cabc T __ZN6reflex14PatternMatcherINS_7PatternEE7patternERKNSt3__112basic_stringIcNS3_11char_traitsIcEENS3_9allocatorIcEEEE
000000000000c994 T __ZN6reflex14PatternMatcherINS_7PatternEE7patternERKS1_
000000000000c0c4 T __ZN6reflex14PatternMatcherINS_7PatternEE7patternERKS2_
...
lib/libreflex.a(pattern.o):
00000000000001e0 T __ZN6reflex11regex_errorC1EiRKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEEm
0000000000012628 T __ZN6reflex11regex_errorD0Ev
0000000000000288 T __ZN6reflex11regex_errorD1Ev
000000000001526c T __ZN6reflex6RangesIjE6insertERKjS3_
0000000000013cc8 T __ZN6reflex6RangesIjEoRERKS1_
0000000000016478 T __ZN6reflex6RangesItE6insertERKtS3_
0000000000016c58 T __ZN6reflex6RangesItEoRERKS1_
00000000000131e4 T __ZN6reflex7ORangesIjEmIERKS1_
000000000000d774 T __ZN6reflex7Pattern10encode_dfaEPNS0_3DFA5StateE
000000000000aa80 T __ZN6reflex7Pattern11analyze_dfaEPNS0_3DFA5StateE
000000000000d540 T __ZN6reflex7Pattern11compact_dfaEPNS0_3DFA5StateE
00000000000011d8 T __ZN6reflex7Pattern12init_optionsEPKc
000000000000a5b0 T __ZN6reflex7Pattern13gen_match_hfaEPNS0_3DFA5StateE
...
```
